### DISCUSSION TITLE
find and replace

### DISCUSSION THREAD

**Discussion by BKSalman** [helix-editor/helix#3630] [Q&A] [answered]
I'm not too good with regex, but is there a way to find and replace using the search("/")?

**Comment by A-Walrus** [comment, marked as answer]
In Helix the usual way to find and replace is to first select with `s`  then  replace with `c`

edit: originally said `r` instead of `c` (`r` replaces with character, with `c` you can type whatever you want)

**Reply by BKSalman** [reply]
I see,
but is there a way to write it out, instead of replacing all with one letter, or replacing with yanked text

**Reply by A-Walrus** [reply]
oops, meant to say replace with `c` :P

**Reply by BKSalman** [reply]
ah :p

but how would I select using ``s`` 
it doesn't work for me{?), or I just don't know how to use it

![image](https://user-images.githubusercontent.com/85521119/187894033-6bd9ce2a-270e-4735-8b22-d46bb650e355.png)

the image is to illustrate that my query isn't being selected

thanks!

**Reply by A-Walrus** [reply]
the `s` select commands takes a subselection of the original. If you want to find and replace within the entire file you must first select the whole file (with `%`), and then use `s` to select the parts that you want, and then replace them with `c`

**Reply by BKSalman** [reply]
I seeeeeee, thanks a lot!

**Reply by hervyqa** [reply]
@A-Walrus 
how about `:%s/foo/bar/gI` in vim?

**Reply by ghost** [reply]
I'd like to search and replace, keeping part of the original selected string.
e.g. in vim : 
`%s/foo\(.*)bar/\t\1 >`
that should change any fooblablablabar into [tab]blablabla >
I am able to find fooblablablabar using '%s' `foo(.*)bar` 
how can I introduce $1 or \1 after typing 'c' ?
thanks

**Reply by anoduck** [reply]
Admittedly, not originating from either vim or kakoune, this bewildered me a heap. When I finally got it to work, it was like pure magic.

**Reply by rayhem** [reply]
I'm still learning Helix so please correct me if I'm wrong, but a big drawback here is how the cursor moves when selecting the whole file. 

# Example workflow

1. I have complex string they want to replace every instance of. The string is nontrivial, so I intend to enter it into the search system by reading it. (It may make sense to yank the string itself, but I may want to add regular expressions or other "decorations" to the search.)
2. I hit % to select the whole file.
3. I am forcibly scrolled to the bottom of the file. The string I was going to read has disappeared from my view.

Ideally there would be a way for me to construct the search-and-replace action without changing anything about the file or where I'm at in it. Vim's `:%s/foo/bar/g` does this naturally; nothing happens until the user enters this _and then_ runs the command.

**Reply by anoduck** [reply]
@rayhem I see what you are talking about. I don't use `:%s/foo/bar/g`, but rather the less adept solution of `substitute.nvim`. Which would be `:substitute/foo/bar/ {N}`. `foo` is what to search for, `bar` is what to replace it with, and `N` is the amount of lines to search for `foo`.

**Reply by dessalines** [reply]
Anyway to make this case sensitive?

**Reply by robrecord** [reply]
> I'd like to search and replace, keeping part of the original selected string. e.g. in vim : `%s/foo\(.*)bar/\t\1 >` that should change any fooblablablabar into [tab]blablabla > I am able to find fooblablablabar using '%s' `foo(.*)bar` how can I introduce $1 or \1 after typing 'c' ? thanks

Currently, it would seem that instead of using `c`, you would edit the selection in such a way as to dance around the part you don't want to change with clever use of movement keys. You could also subselect using `s`, or sub-deselect a part of the selection with `alt-K` (remove_selections) and edit the rest, if your edit will be symmetrical. Or you can just pipe whatever selection to `sed` with `|` (shell_pipe).

For your example, you could do:
- `%` to select all
- `s` to select with regex: `foo.*bar` then `<enter>`
- `s` to select with regex: `(foo|bar)` then `<enter>`. This selects foo or bar within the selection.
- `d` to delete selection.

**Reply by robrecord** [reply]
> Anyway to make this case sensitive?

Yes, 
- if you have used any capital letters in the query, or
- if you prefixed your query with `(?-i)`

See https://github.com/helix-editor/helix/issues/11981

**Reply by dessalines** [reply]
Sweet, thx.

**Reply by michihupf** [reply]
did you find a way to define and use regex groups for replacement?

**Comment by rickswe** [comment]
The only problem is there's no feedback like in vim with `gc`. You basically have to trust the
multiple cursors or "you" the user to input properly. Not sure if the muliple cursors was the
best approach to this. Probably a better solution would have been to make a dedicated
command or function but why reinvent the wheel when multiple cursors are already
part of it? ...

```text
:%s/Foo/Bar/gc
```

**Comment by blitzarx1** [comment]
And what about find and replace in all project files?

**Reply by kirawi** [reply]
There was an abandoned PR for this: https://github.com/helix-editor/helix/pull/4381

However, if you are using an LSP you can use space-r

**Reply by rickswe** [reply]
A vimgrep clone would be great but I think right now there are more important things that need attention in helix. For now sed will do the trick.

**Comment by stesj4** [comment]
I want to find and replace, but not with multi cursor. In my document I don't want replace all hits. Now I search and find, exit multi cursor and use the n key to jump and edit them individually. No replace

**Reply by rossmeier** [reply]
Found a great way to do this today:
- Search using `/`
- On first occurrence, press `c` and replace with what you want to replace
- A: press `n` to go to next occurrence
- If you want to replace this occurrence, press `.` to replace with last inserted (i.e. what you replaced the last thing with)
- goto A

**Reply by stesj4** [reply]
Thank you very much, just what I was looking for 😊

**Reply by lihsai0** [reply]
> Found a great way to do this today:
> 
> * Search using `/`
> * On first occurrence, press `c` and replace with what you want to replace
> * A: press `n` to go to next occurrence
> * If you want to replace this occurrence, press `.` to replace with last inserted (i.e. what you replaced the last thing with)
> * goto A

@rossmeier In vim, a better way to do it is search first, then `cgn`, and here just use only one key `.` to replace & find next. I'm finding how to do this in helix, but have no idea. T^T

**Reply by gyreas** [reply]
> Found a great way to do this today:
> - Search using `/`
> - On first occurrence, press `c` and replace with what you want to replace
> - A: press `n` to go to next occurrence
> - If you want to replace this occurrence, press `.` to replace with last inserted (i.e. what you replaced the last thing with)
> - goto A

I literally came to add this trick. Seems you called dips on it :D

**Comment by seungjin** [comment]
Multicursor way works fine but had a trouble when I have multiple same thing to replace on single lines.
For example, like to replace all \t to 4 spaces with 
```
\tHello
\t\t Hello World
\t\t\t Bye
``` 
Simply I do : 
```
%
:pipe sed 's/\t/   /g'
````

**Comment by langdine** [comment]
I can't understand why Helix doesn't support a substitute command. Having to pipe text to another process just to perform such a simple replacement feels unnecessary. Pipe to "sd" have a nice work.

**Reply by michihupf** [reply]
it's one of the main reasons I switched back to nvim last week. I see that most people might be fine using multiple cursors, but for more complex replacements this just sucks. I might revisit helix later sometime if they someday fix this
